tiatoolbox/models/architecture/vanilla.py

```diff
@@ -0,0 +1,160 @@
+    elif "densenet" in arch_name:
+        feat_extract = model.features
+    elif "alexnet" in arch_name:
+        feat_extract = model.features
+    if "inception_v3" in arch_name or "googlenet" in arch_name:
+        feat_extract = nn.Sequential(*list(model.children())[:-3])
+    if "mobilenet" in arch_name:
+        feat_extract = model.features
+    return feat_extract
+
+
+class CNNModel(ModelABC):
+    """Retrieve the model backbone and attach an extra FCN to perform classification.
+
+    Args:
+        backbone (str): Model name.
+        num_classes (int): Number of classes output by model.
+
+    Attributes:
+        num_classes (int): Number of classes output by the model.
+        feat_extract (nn.Module): Backbone CNN model.
+        pool (nn.Module): Type of pooling applied after feature extraction.
+        classifier (nn.Module): Linear classifier module used to map the features
+          to the output.
+
+    """
+
+    def __init__(self, backbone, num_classes=1):
+        super().__init__()
+        self.num_classes = num_classes
+
+        self.feat_extract = _get_architecture(backbone)
+        self.pool = nn.AdaptiveAvgPool2d((1, 1))
+
+        # Best way to retrieve channel dynamically is passing a small forward pass
+        prev_num_ch = self.feat_extract(torch.rand([2, 3, 96, 96])).shape[1]
+        self.classifer = nn.Linear(prev_num_ch, num_classes)
+
+    # pylint: disable=W0221
+    # because abc is generic, this is actual definition
+    def forward(self, imgs):
+        """Pass input data through the model.
+
+        Args:
+            imgs (torch.Tensor): Model input.
+
+        """
+        feat = self.feat_extract(imgs)
+        gap_feat = self.pool(feat)
+        gap_feat = torch.flatten(gap_feat, 1)
+        logit = self.classifer(gap_feat)
+        prob = torch.softmax(logit, -1)
+        return prob
+
+    @staticmethod
+    def postproc(image):
+        """Define the post-processing of this class of model.
+
+        This simply applies argmax along last axis of the input.
+        """
+        return np.argmax(image, axis=-1)
+
+    @staticmethod
+    def infer_batch(model, batch_data, on_gpu):
+        """Run inference on an input batch.
+
+        Contains logic for forward operation as well as i/o aggregation.
+
+        Args:
+            model (nn.Module): PyTorch defined model.
+            batch_data (ndarray): A batch of data generated by
+              torch.utils.data.DataLoader.
+            on_gpu (bool): Whether to run inference on a GPU.
+
+        """
+        device = misc.select_device(on_gpu)
+
+        img_patches = batch_data
+        img_patches_device = img_patches.to(device).type(torch.float32)  # to NCHW
+        img_patches_device = img_patches_device.permute(0, 3, 1, 2).contiguous()
+
+        # Inference mode
+        model.eval()
+        # Do not compute the gradient (not training)
+        with torch.no_grad():
```

Comment:
There was a recent tweet from torch about a faster alternative to `torch.no_grad()` at inference. I will try to find it an link to it here.

Reply:
Tweet: https://twitter.com/PyTorch/status/1437838231505096708?s=20
Docs: https://pytorch.org/docs/1.9.0/notes/autograd.html#locally-disabling-gradient-computation